src/units/abilities.cpp

```diff
@@ -1258,16 +1258,19 @@ unit_ability_list attack_type::get_weapon_ability(const std::string& ability) co
 unit_ability_list attack_type::get_specials_and_abilities(const std::string& special) const
 {
 	unit_ability_list abil_list = get_weapon_ability(special);
-	if(!abil_list.empty()){
-		abil_list = overwrite_special_checking(abil_list, abil_list, false);
-	}
 	unit_ability_list spe_list = get_specials(special);
-	if(!spe_list.empty()){
-		spe_list = overwrite_special_checking(spe_list, abil_list, true);
-		if(special == "plague" && !spe_list.empty()){
+	abil_list.append(spe_list);
+	unit_ability_list overwriters = overwrite_special_overwriter(abil_list);
+	if(!abil_list.empty() && !overwriters.empty()){
+		abil_list = overwrite_special_checking(abil_list, overwriters);
+	}
+	if(special == "plague" && !spe_list.empty()){
```

Comment:
I've tried simplifying this function, and it passes the unit tests but I'm not sure if I'm missing something.

IIUC, some callers only look at the first item on the list, and this special-case for `plague` is to ensure that a weapon special will be returned before an ability used as a weapon special. I think the entire special case code can be removed by calling `get_specials()` before `get_weapon_ability()`:

```
unit_ability_list attack_type::get_specials_and_abilities(const std::string& special) const
{
	unit_ability_list abil_list = get_specials(special);
	abil_list.append(get_weapon_ability(special));
	unit_ability_list overwriters = overwrite_special_overwriter(abil_list);
	if(!abil_list.empty() && !overwriters.empty()){
		abil_list = overwrite_special_checking(abil_list, overwriters);
	}
	return abil_list;
}
```

Reply:
Ok, i will try that